File: Leetcode/Graphs/785. Is Graph Bipartite?.cpp
class Solution {
private:
    bool DFS(int source, vector <int> &Color, vector<vector<int>>& graph, int color) {
        if (Color[source] == -1) {
            Color[source] = color;
            bool valid = true;
            int M = graph[source].size();
            for (int i = 0; i < M; i++) {
                if (!(DFS(graph[source][i], Color, graph, 1 - color))) {
                    valid = false;
                    break;
                }
            }
            return valid;
        } else return Color[source] == color;
    }
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int N = graph.size();
        bool isGraphBipartite = true;
        vector <int> Color (N, -1);
        for (int i = 0; i < N; i++) {
            if (Color[i] == -1) {
                if (!(DFS(i, Color, graph, 1))) {
                    isGraphBipartite = false;
                    break;
                }
            }
        }
        return isGraphBipartite;
    }
};

File: Leetcode/Graphs/785. Is Graph Bipartite?.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        Color = [-1 for i in range(len(graph))]

        def DFS(source, Color, graph, color):
            if Color[source] == -1:
                Color[source] = color
                M, valid = len(graph[source]), True
                for i in range(M):
                    if not DFS(graph[source][i], Color, graph, 1 - color):
                        return False
            return Color[source] == color

        for i in range(len(graph)):
            if Color[i] == -1:
                if not DFS(i, Color, graph, 1):
                    return False
        return True
